### PR TITLE
Draft1

### PR DESCRIPTION
**Pull request by fracasula**
## Description of the change

First draft.

Before reviewing, have a look at the [README.md](https://github.com/rudderlabs/rudder-load/blob/draft1/README.md) to understand how it works.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+COPY cmd cmd
+COPY internal internal
+COPY samples samples
+
+# Build binary
+RUN CGO_ENABLED=0 GOOS=linux go build -a -installsuffix -ldflags="-s -w" \
+    -o ./rudder-load-producer \
+    cmd/producer/main.go
+
+FROM ubuntu:noble
```
Comment (lvrach) [inline Dockerfile]: [minor] why not alpine ?

File: samples/page.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{
+  "type": "page",
+  "name": "Home",
+  "properties": {
+    "title": "Home | RudderStack",
+    "url": "https://www.rudderstack.com"
+  },
+  "messageId": "<MSG_ID>",
```
Comment (lvrach) [inline samples/page.json]: can we use templating for this ?

File: cmd/producer/main.go [hunk 3]
```diff
@@ -0,0 +1,693 @@
...
+		collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}),
+	)
+	publishRatePerSecond := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "publish_rate_per_second",
+		Help: "Publish rate per second",
+		ConstLabels: map[string]string{
+			"mode":       mode,
+			"slots":      strconv.Itoa(len(slots)),
+			"total_keys": strconv.Itoa(totalKeys),
+		},
```
Comment (lvrach) [inline cmd/producer/main.go]: Can you  explain those labels?

File: README.md [hunk 4]
```diff
@@ -1 +1,54 @@
...
+This repository is used to generate an artificial load to test the RudderStack architecture under intense traffic.
+
+## Usage
+
+The service is designed to be run in a Kubernetes cluster.
+
+It is possible to create multiple replicas of the service to increase the load.
+
+Each replica owns a specific writeKey and will send events via multiple go routines. Each go routine will send events
+owning a set of keys. This way we can guarantee in-order delivery of events per writeKey-userId.
```
Comment (lvrach) [inline README.md:12]: Is event ordering a requirement for load tester ?

File: README.md [hunk 5]
```diff
@@ -1 +1,54 @@
...
+
+## Example configuration
+
+1. Replicas: 3
+2. WriteKeys: 3
+3. Concurrency: 200 (number of go routines per replica, we'll call them **slots**)
+4. Message generators: 10
+   a. number of go routines that will generate events and send them to the **slots** (see point 3)
+5. Keys per slot map: 50
```
Comment (lvrach) [inline README.md]: What is a key ? writeKey or userId ?

File: README.md [hunk 6]
```diff
@@ -1 +1,54 @@
...
+   a. number of **unique** keys that each slot will own
+   b. this can be a map as well like 10_20_30_40 which would mean that 50 slots will own 10 keys, 
+      50 slots will own 20 keys, 50 slots 30 keys and 50 slots 40 keys (because we create 4 groups,
+      and we have 200 slots)
+6. Traffic distribution percentage: 60_40
```
Comment (lvrach) [inline README.md]: how can we define the distribution for diffrent event types ?

File: Dockerfile (hunk 1 above)
Comment (fracasula) [reply Dockerfile]: it was just to be able to copy .deb and other deps for debugging. I'm not using it anymore though so we can use alpine :+1:

File: cmd/producer/main.go (hunk 3 above)
Comment (fracasula) [reply cmd/producer/main.go]: Done.

File: README.md (hunk 4 above)
Comment (fracasula) [reply README.md:12]: If you need it is there. If you don't need it just pass the same write key over and over :man_shrugging:

File: README.md (hunk 5 above)
Comment (fracasula) [reply README.md]: If you `RANDOM_KEY_NAMES: "true"` they will be random.

If you don't do that then they'll look like this:

```
writeKey+"-key-"+strconv.Itoa(i)+"-"+strconv.Itoa(j)
```

where `i` is the `slot` publishing and `j` is the index of the key. if you have 50 keys per slot then `j` will range from `0` to `49`. this makes them predictable even if you re-run the script multiple times.

Comment (fracasula) [reply README.md]: I added this to the `README.md`.

File: README.md (hunk 6 above)
Comment (fracasula) [reply README.md]: You can't. Now you can either send a `page` or a `batch` with inside the same `page` over and over (only `messageId` will change).

**Comment by fracasula** [comment]
@lvrach feel free to close this one in favour of https://github.com/rudderlabs/rudder-load/pull/2

**Pull request closed without merging**